openedx/core/djangoapps/import_export/courses/views.py

```diff
@@ -0,0 +1,530 @@
+from xmodule.modulestore import COURSE_ROOT, LIBRARY_ROOT
+
+from student.auth import has_course_author_access
+
+from openedx.core.lib.extract_tar import safetar_extractall
+from util.json_request import JsonResponse
+from util.views import ensure_valid_course_key
+
+from urllib import urlencode
+
+log = logging.getLogger(__name__)
+
+
+# Regex to capture Content-Range header ranges.
+CONTENT_RE = re.compile(
+    r"(?P<start>\d{1,11})-(?P<stop>\d{1,11})/(?P<end>\d{1,11})"
+)
+
+
+class IsCourseStaff(BasePermission):
+    """
+    Permission that checks to see if the request user has permission to access
+    all course content of the requested course
+    """
+    def has_permission(self, request, view):
+        course_key_string = view.kwargs['course_key_string']
+        try:
+            course_key = CourseKey.from_string(course_key_string)
+        except InvalidKeyError:
+            raise Http404
+
+        return has_course_author_access(request.user, course_key)
+
+
+class ArchiveRenderer(renderers.BaseRenderer):
+    """
+    A Renderer for compressed tars. It gets used at the content negotiation
+    stage, but "render" never actually gets used.
+    """
+    media_type = "application/x-tgz"
+    format = None
+    render_style = "binary"
+
+    def render(self, data, _media_type=None, _render_context=None):
+        return data
+
+
+class FullCourseImportStatus(APIView):
+    """
+    View the import status of a full course import.
+    """
+    authentication_classes = (OAuth2Authentication, SessionAuthentication)
+    permission_classes = (IsAuthenticated, IsCourseStaff)
+
+    @ensure_valid_course_key
+    def get(self, request, course_key_string, filename=None):
+        """
+        Returns an integer corresponding to the status of a file import.
+        These are:
+
+            -X : Import unsuccessful due to some error with X as stage [0-3]
+            0 : No status info found (import done or upload still in progress)
+            1 : Extracting file
+            2 : Validating.
+            3 : Importing to mongo
+            4 : Import successful
+
+        """
+        status_key = "{}|{}{}".format(
+            request.user.username,
+            course_key_string,
+            filename
+        )
+        status = cache.get(status_key, 0)
+
+        return Response({"ImportStatus": status})
+
+
+class FullCourseImportExport(APIView):
+    """
+    Import or export a full course archive.
+    """
+    authentication_classes = (OAuth2Authentication, SessionAuthentication)
+    permission_classes = (IsAuthenticated, IsCourseStaff)
+    renderer_classes = (ArchiveRenderer, JSONRenderer)
+
+    def _save_request_status(self, request, key, status):
+        """
+        Save import status for a course in request session
+        """
+        cache.set(
+            "{}|{}".format(request.user.username, key),
+            status
+        )
```

Comment:
The above diff reference is drifting away from it's initial position due to changes I've been making to address other comments. The lines I'm talking about are:
```python
except Exception as exception: # pylint: disable=broad-except
    self._save_request_status(request, courselike_string, -1)
```